photon-api/src/main/scala/com/linkedin/photon/ml/optimization/game/CoordinateOptimizationConfiguration.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.optimization.game
+
+import com.linkedin.photon.ml.optimization._
+
+/**
+ * Generic trait for a configuration to define a coordinate.
+ */
+sealed trait CoordinateOptimizationConfiguration
```

Comment:
@joshvfleming 
@fastier-li 
@Adama94

A new trait to unify the old `GLMOptimizationConfiguration` and `FactoredRandomEffectOptimizationConfiguration`. The down-sampling rate has been pulled out of `GLMOptimizationConfiguration`, which has also been extended by two derived classes `FixedEffectOptimizationConfiguration` and `RandomEffectOptimizationConfiguration`. This is used to enforce a stricter type safety throughout the code, so that the library code expects the coordinate type of data and optimization configurations to match.